Tahi-project/tahi
# APERTA-7248: retain word docs

JIRA issue: [APERTA-7248](https://developer.plos.org/jira/browse/APERTA-7248#)
### What this PR does:

This changeset keeps the history of all uploaded manuscripts. This is to allow a developer or sysadmin to be able to revert/recover a previous version of the manuscript.

This changeset does include set of db/migrations
### Notes
- This introduces `paper-trail` for handling keeping track of old paper versions.
- This moves the manuscript from `VersionedText#source` to a new model `ManuscriptAttachment`. 
  - This does NOT alter `VersionedText` in any other way.
- This preserves history of existing MS history that Aperta knows about
  - It does NOT move existing MS files on S3, it leaves them where they are.
- New MS files will be stored based on the `AttachmentUploader` storage rules. So new MS files will be stored with a file hash, whereas existing ones will not.
### Major UI changes

None. There are no UI changes.
### How to test

Here's how to set things up. 
- First, get on `master`
- Reset your database
- Load the app.
- Create a new paper (in `rails console` the paper should have 1 `versioned_texts` record).
- Submit the paper for submission
- Register a decision
- Upload a new manuscript (using the Upload Manuscript card). 
- In `rails console` Your paper should have 2 `versioned_texts`.
- Write down the results of:
  - `VersionedText.count`
  - `Paper.all.map { |p| p.latest_version }.count`

Now, kill the app, and switch to this (`features/APERTA-7248_retain-uploaded-word-doc`) branch, and run `rake db:migrate`. Load up `rails console` and you should see the following:
- The count of `VersionedText` records should not have been affected. 
- The count of `ManuscriptAttachment` records should match the number of papers that had a `latest_version`
- If you load up your paper you should see that `paper.file` is a `ManuscriptAttachment` that points to the docx file that you most recently uploaded to it.
- `paper.file.versions` should contain the history of your paper. 

Now, if load up the app and log in thru the UI, you should be able to go back to your paper and any other paper and go to `Downloads > WORD` and be able to download the manuscript. This should still work.

Next, create a new paper in Aperta, and do at least one round of decisions, and upload a new MS after each decision:
- In the UI, you should still be able to download the latest DOCX thru `Downloads > WORD`.
- In `rails console` you should see that there is every only 1 ManuscriptAttachment for a paper, but you should see that as you upload new manuscripts the `paper.file.versions` list gets longer.
#### Now, let's perform a recovery:

Make sure you have workers, slanger, and ihat all running. Load up your browser and go to the MS page for a paper that you want to recover. 

Next, load up `rails console` and perform the following steps:
- Load up your paper and load up the history, e.g. `paper.file.versions`. 
- Select a version: `version = paper.file.versions.last`
- Re-ify the version: `model = version.reify` (This doesn't persist the change, just makes the version an `ManuscriptAttachment` instance)
- Now restore the reified version by saving it: `model.save`
- Note: you can also do this by assigning the the reified attachment to `Paper#file`, e.g. `paper.file = paper.file.versions.last.reify` because this will automatically issue a save.

Note: if you restore a previous version that had pointed to the same file than _nothing will happen_. 

When you save a restored version it should fire up background jobs and re-process the document in ihat and then it should automatically update the MS page so you see that the text has been updated.
#### Why re-send ihat jobs?

We have the recovery process work thru the same mechanism that normal MS/docx uploads go thru. This is because there are secondary actions that need to take place in order to process a paper (such as figure processing/placement, title and abstract extract, etc) .

We don't want to have to code in special cases for each of these so the restore process acts like a new upload.
#### Animated GIF of restoring an old version

You may need to view this in Chrome. Safari had trouble playing it for me (Zach).

![aperta-7248-restore](https://cloud.githubusercontent.com/assets/967/18142133/36ae38ba-6f8a-11e6-9d19-ffbe9bbddcef.gif)

---
### Code Review Tasks:

Author tasks:

If I modified any environment variables:
- [ ] I made a pull request to change the files on the [molten repo](https://github.com/PLOS/molten/tree/dev/pillar/aperta) {PR LINK}
- [ ] I double-checked the `app.json` file to make sure that the heroku review apps are still inheriting the correct environment variables from staging

Reviewer tasks:
### After the Code Review:

Author tasks:
